### PR TITLE
#162948977 Write unit tests for ALL existing functionality

### PR DESCRIPTION
**Pull request by JoshuaKodhe**
#### What does this PR do?

Have tests for the existing functionality

#### Description of Task to be completed?

Have tests for the Login functionality 
Have tests for the Signup functionality 
Have tests for the User model 

#### How should this be manually tested?
After cloning the repo using `git clone https://github.com/andela/ah-fulldeck.git`
Move into the project folder `cd <path_of_the_project>`
Check out this branch `ch-test-existing-functionality-162948977` using the command
`git checkout ch-test-existing-functionality-162948977`
Run the tests using the command `python manage.py test`

#### What are the relevant pivotal tracker stories?

[#162948977 ](https://www.pivotaltracker.com/story/show/162948977)

#### Screenshots 
screenshot 2019-01-15 at 15 18 18

### CODE REVIEW

**Comment by Nicanor008** [comment]
I like how you have done your tests, am getting some ideas on how I'll implement my task from your files.

**Comment by Hesbon5600** [comment]
I like the idea of having a separate file for the base test.  
Great implementation of tests for the  models  💪

File: authors/apps/authentication/tests/test_login.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    def test_no_email_login(self):
+        self.registration
+        del self.test_user['user']['username']
+        del self.test_user['user']['email']
+        response = self.client.post(self.login_url,
+                                    self.test_user,
+                                    format='json')
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertIn(b'This field is require', response.content)
```
Comment (eduhmik) [inline authors/apps/authentication/tests/test_login.py]: Kindly amend the message in assertIn

File: authors/apps/authentication/tests/test_login.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    def test_no_password_login(self):
+        self.registration
+        del self.test_user['user']['username']
+        del self.test_user['user']['password']
+        response = self.client.post(self.login_url,
+                                    self.test_user,
+                                    format='json')
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertIn(b'This field is require', response.content)
```
Comment (eduhmik) [inline authors/apps/authentication/tests/test_login.py]: Here too

**Comment by charitymarani** [comment]
@JoshuaKodhe Could you test unregistered user trying to log in,   I mean a non-existent account

**Review by kzyangiro — CHANGES_REQUESTED** [review]
@JoshuaKodhe Kindly clear the Base app and move the welcome message to auth, then point the testing to all remaining apps, not only Authentication

File: authors/apps/authentication/tests/test_signup.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertIn(b'This field is required', response.content)
+
+    def test_no_password_registration(self):
+        del self.test_user['user']['password']
+        response = self.client.post(self.signup_url,
+                                    self.test_user,
+                                    format='json')
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertIn(b'This field is required', response.content)
```
Comment (kzyangiro) [inline authors/apps/authentication/tests/test_signup.py]: @JoshuaKodhe This chunk of code is duplicated in about 4 or 5 more places. We need to be very keen on maintainability, create a single function/method that does this, then call it in all the relevant areas to reduce code duplication

File: authors/urls.py [hunk 4]
```diff
@@ -24,6 +24,6 @@
     path('admin/', admin.site.urls),
 
     path('api/v1/', include('authors.apps.authentication.urls',
-                            namespace='authentication')),
-    path('', include('authors.apps.base.urls')),
+                            namespace='app_authentication')),
+    path('', include('authors.apps.authentication.urls')),
```
Comment (kzyangiro) [inline authors/urls.py:28]: @JoshuaKodhe Why did you name it app_authentication rather than just authentication?

**Review by kzyangiro — APPROVED** [review]

**Pull request merged**